sphinx_doc/tutorials/docs/writing_new_costmap2d_plugin.rst

```diff
@@ -0,0 +1,240 @@
+
+The plugin's main class ``nav2_gradient_costmap_plugin::GradientLayer`` is inherited from basic class ``nav2_costmap_2d::Layer``:
+
+.. code-block:: c
+
+  namespace nav2_gradient_costmap_plugin
+  {
+  
+  class GradientLayer : public nav2_costmap_2d::Layer
+
+The basic class provides the set of virtual methods API for working with costmap layers in a plugin. These methods are calling in runtime by ``LayeredCostmap``. The list of methods, their description and necessity to have these methods in plugin's code is presented in the table below:
+
++----------------------+----------------------------------------------------------------------------+-------------------------+
+| **Virtual method**   | **Method description**                                                     | **Requires override?**  |
++----------------------+----------------------------------------------------------------------------+-------------------------+
+| onInitialize()       | Method is called at the end of plugin initialization. There is usually     | No                      |
+|                      | declarations of ROS parameters with its default values.                    |                         |
++----------------------+----------------------------------------------------------------------------+-------------------------+
+| updateBounds()       | Method is called to ask the plugin: which area of costmap layer it needs   | Yes                     |
+|                      | to update. There are 3 input parameters of method: robot position and      |                         |
+|                      | orientation and 4 output parameters: pointers to window bounds.            |                         |
+|                      | These bounds are used for performance reasons: to update the area          |                         |
+|                      | inside window where is new info available, avoiding updates of whole       |                         |
+|                      | costmap on every iteration.                                                |                         |
+|                      |                                                                            |                         |
+|                      |                                                                            |                         |
++----------------------+----------------------------------------------------------------------------+-------------------------+
+| updateCosts()        | Method is called each time when costmap re-calculation is required. It     | Yes                     |
+|                      | updates the costmap layer only within its bounds window. There are 4 input |                         |
+|                      | parameters of method: calculation window bounds and 1 output parameter:    |                         |
```

Comment:
"local costmap" means something else, so I wouldn't use that term. How about:

```
The ``Layer`` class provides the plugin with an internal costmap, ``costmap_``, for updates. The ``master_grid`` should be updated with values within the window bounds using one of the following update methods: ... 
```

Reply:
Done.